add todo (need review)

adrienluitot opened the pull request.

Commit:
add todo (need review)

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+from settings import *
+from checks import *
+
+TOKEN = "NTY4MDkyNTY5NTE5NzgzOTM2.XMiBDw.zCL0KhV_gOtxfyACctgYCLcLlEY"
```
@adrienluitot you should delete this token

NotaSmartDev replied:
This line should be deleted before I can merge...

s0lst1ce replied:
This line should be deleted before I can merge...

adrienluitot replied:
Already done ;)

s0lst1ce replied:
Why am I not seeing it then ? You haven't you committed ?

adrienluitot replied:
it's that

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+    async def on_message(self, message):
+        if message.author == bot.user:
+            return
+        await bot.process_commands(message)
```
you don't need to do this in a cog, not hte way the bot is built. So you can remove the listener.

s0lst1ce replied:
Same here, as you agreed I need you to delete this listener.

adrienluitot replied:
I didn't psh yet but done

s0lst1ce replied:
Oh I see. Well please commit asap. It would prevent me from writing useless comments ;)

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+        if reaction.user_id != bot.user.id:
+            message = await bot.get_channel(reaction.channel_id).fetch_message(reaction.message_id)
+
+            with open(TODO_CHANNEL, "r") as file:
```
is it a channel or a file ? This seems unclear to me...

s0lst1ce replied:
Also where do you define this variable ?

adrienluitot replied:
It's an error from, I corrected that, It's a file I also fixed that

s0lst1ce replied:
Ok I'll wait for the commit then ;)

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+            message = await bot.get_channel(reaction.channel_id).fetch_message(reaction.message_id)
+
+            with open(TODO_CHANNEL, "r") as file:
+                channelId = file.readlines()[0]
```
you only need the first line so instead of reading the whole file you can just do `readline` which will only read one line with each iteration.

adrienluitot replied:
Thanks, will fix that

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+
+                        repostFieldValue = None
+                        for field in message.embeds[0].fields:
+                            if field.name == "Public repost":
```
instead of having a single allowed name "Public repost" have global var so that is can be configured once I implement #4

adrienluitot replied:
What's the goal ?

s0lst1ce replied:
one day the bot might extend its reach to several servers so I want to keep it ready for such an eventuality.

s0lst1ce replied:
As we now know that the bot is going to be used in more than one server, I'll  be waiting for you to fix this.

adrienluitot replied:
I made a PUBLIC_REPOST constant, even if I'm not sure if it was what you wanetd

s0lst1ce replied:
:man_facepalming: Sry I'm such an idiot. I read through your code a little too fast it would seem. I completely missed misunderstood the use of this thing. It probably wasn't needed. Sry I made a big mistake here :grimacing:

adrienluitot replied:
Oh, so I remove it ? We can also let it in memory to make translation xD

s0lst1ce replied:
Do whatever you want I don't have my word in this matter anymore...

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+                    elif reaction.emoji.name == EMOJIS['hourglass']:
+                        await message.add_reaction(EMOJIS['check'])
+
+    @commands.group()
```
shouldn't the use of this group of commands be restricted ?

adrienluitot replied:
Yes ! It should :D

s0lst1ce replied:
Maybe you should add a new clearance level that could be configured ( see #4 ) ?

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+    @todo.command()
+    async def removetype(self, ctx, command):
+        '''Command to add a remove type.'''
+        with open(TODO_TYPES, "r") as file:
```
Where is `TODO_TYPES` defined at ? Your PR doesn't include the settings.py file so I'm guessing you didn't change it ? I'm a bit lost tbh.

adrienluitot replied:
I didn't push the changes but I added it (in settings.py)

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+        '''Command to add a todo. Usage : <the thing todo>;<type of todo>;<assigned to / false>;<repost public channel / false>'''
+        with open(TODO_CHANNEL, "r") as file:
+            channelId = file.readlines()[0]
+            if channelId == None or channelId == "":
```
camel case shouldn't be used in python. Ever. If you'r referencing classes you can do something like `ChannelName`. Functions and variables should look like: `channel_name`. See PEP8 for more info.

adrienluitot replied:
Oh ! okay I confused camel case and snake case >.< sorry !

s0lst1ce replied:
np

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+
+        command = ""
+        for arg in args:
+            command += " {}".format(arg)
```
What is the point of the space in `" {}"` ?

adrienluitot replied:
In fact I think I had copy your code x) and It's probably to keep spaces `(to keep "make a todo list" instead of "makeatodolist")` :smile:

s0lst1ce replied:
Ok

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+        command = ""
+        for arg in args:
+            command += " {}".format(arg)
+        command = command.split(";")
```
Maybe letting users issue commands with multiple args only using spaces would make it more natural to them. This would make it closer to how the CLI works as well as most other bots on discord. It also wouldn't cause any additional difficulty and would probably be clearer.

adrienluitot replied:
Hum, the problem is that I need to keep spaces for todo's descriptions. Otherwise they have to use quotes.

s0lst1ce replied:
That is true. Although a way to fix this would be to put the message at the end. Like this all words after the commands would be part of the message. What do you think ?

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+            command += " {}".format(arg)
+        command = command.split(";")
+
+        todoType = None
```
Again, camel case

adrienluitot replied:
Changed it ;)

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+        else:
+            embedColor = 0x28a745
+        
+        newEmbed = discord.Embed(description=command[0], url="", color=embedColor)
```
Again, camel case

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+        todoType = None
+
+        with open(TODO_TYPES, "r") as file:
+            lines = file.readlines()
```
I think I already mentioned it but as I'm seeing it again: I think using `readline` would be more efficient.

adrienluitot replied:
Yep, done but I didn't push yet

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+        
+        command[2] = command[2].replace(' ', '') #TODO: Use dfind instead ?
+        if command[2] != "false":
+            if command[2].startswith("<@"):
```
I used to cut the string this way too but it's most unefficient. Look at my updated code instead. If you don't see where I used to do this and how I solved this please ping me and I'll do my best to show you how to use the API to its full extent.

adrienluitot replied:
Okay I'll take a look

adrienluitot replied:
Is this only `member:discord.Member` in the parameters of the function ? Else I don't see what you done

s0lst1ce replied:
I checked and that seems to be it. I didn't want to say that was it for sometimes I forget part of my code ^^

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+        command[2] = command[2].replace(' ', '') #TODO: Use dfind instead ?
+        if command[2] != "false":
+            if command[2].startswith("<@"):
+                user = ctx.guild.get_member(int(command[2][2:-1]))
```
Maybe you could use a `discord.py` converter here ? Not sure if it'd be better, I'll let you judge about it but keep me informed ;)

adrienluitot replied:
How do you do ? ^^

s0lst1ce replied:
Look at the doc :P

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+            else:
+                for chan in ctx.guild.channels:
+                    if(chan.name == command[3]):
+                        repostChannel = chan
```
Again, camel case
```suggestion
                        repost_channel = chan
```

adrienluitot replied:
Already done in my code, sorry to avoid problem I don't commit what you did :open_mouth:

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+                user = ctx.guild.get_member_named(command[2])
+            
+            if user != None:
+                newEmbed.add_field(name="Asssigned to", value=user.mention, inline=True)
```
Again, camel case

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+                newEmbed.add_field(name="Asssigned to", value=user.mention, inline=True)
+        
+        if command[3] != "false":
+            if command[3].startswith("<#"):
```
Again, you don't need to cut your strings in pieces like this.

adrienluitot replied:
How would you do so ? :smiley:

s0lst1ce replied:
Again using a converter.

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+        newEmbed.set_footer(text=command[1])
+        if repostChannel != None:
+            publicTodo = await repostChannel.send(embed=newEmbed)
+            newEmbed.add_field(name="Public repost", value=repostChannel.mention + " : " + str(publicTodo.id), inline=True)
```
Again, camel case. And I'll stop writing this every time I see camel case again. I'll let you review you own code ;)

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+            command[1] = command[1][1:]
+        if len(command[1]) != 6:
+            if len(command[1]) != 3:
+                await ctx.send('The color must be in hexadecimal, like this **#ccc** or **#ff0000**')
```
Should we add more "natural" names for the colors ? This could be useful not only for this project but for other commands too. May even save us, the coders some time ^^

adrienluitot replied:
Hmm yeah not bad idea but, then how ? users will use text like "red" ? 
Also there is some basic colors already done see https://discordpy.readthedocs.io/en/latest/api.html#discord.Colour

s0lst1ce replied:
oh I didn't know. Well this may reveal itself useful one day

s0lst1ce commented on Todo.py:
```diff
@@ -0,0 +1,203 @@
+        await ctx.channel.send('Okay ! This channel wil be used for the todos !')
+
+def setup(bot):
+    bot.add_cog(Todo(bot))
```
It is recommended to add a newline at the end of a file.
```suggestion
    bot.add_cog(Todo(bot))
    
```

adrienluitot replied:
Same as your last change :D

s0lst1ce reviewed (changes requested):
A few minor modifications should be commited before I can merge this PR into `rewrite`. You've done a great job so I think you'll be able to solve all of those issues by yourself but if you do need/want help don't hesitate to ping me. I'll try and be available although I don't code so much :(

Commit:
improved the todo bot and fixed some bugs

also added gitignore
(Corrections and enhancements are all based on @NotaSmartDev 's comment :D)

NotaSmartDev commented:
I'm going to merge your PR and then you'll be able to switch to converters after-wise but there's no reason to withhold the merge into `rewrite` now.
